### PR TITLE
Begin implementation of wasi-http

### PR DESCRIPTION
**Pull request by brendandburns**
Initial draft, working, but with limitations.

Known issues:
* Outbound requests only, inbound HTTP serving is not supported
* ~Body for requests is not supported~
* ~Request options (e.g. timeout values) are not supported.~
* ~There are panics all over the place if you pass in bad pointers or unexpected values.~

### CODE REVIEW

**Comment by brendandburns** [comment]
@pchickey I think this is ready for an initial high-level review.

Everything is implemented in terms of the `wit-bindgen` interfaces, and then an adapter is introduced which maps from the wit-bindgen to the current `wasmtime::Linker` implementation.

This means that it is very forward compatible with the component model linker going forward.

If you can take a quick look at the rough layout and see if this is on the path to merging that would be great.

After we get the rough layout right, we can move on to a more detailed review.

Thanks!

**Comment by brendandburns** [comment]
There's also a pretty extensive example that can be used to test this implementation in `crates/wasi-http/example`

**Review by pchickey — COMMENTED** [review]
Exciting! Here's my first pass of feedback

File: crates/wasi-http/src/component_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/component_impl.rs:85]: The compiler doesn't like this function b/c of Rust borrow semantics when I write it as a function:

```
error[E0382]: borrow of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:343:37
    |
330 |         move |mut caller: Caller<'_, T>, stream: u32, len: u64, ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
339 |             let out_ptr = allocate_guest_pointer(caller, body_len);
    |                                                  ------ value moved here
...
343 |             let memory = memory_get(&mut caller).unwrap();
    |                                     ^^^^^^^^^^^ value borrowed here after move

error[E0382]: use of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:390:55
    |
377 |         move |mut caller: Caller<'_, T>, fields: u32, out_ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
382 |             let tuple_ptr = allocate_guest_pointer(caller, (16 * header_len).try_into().unwrap());
    |                                                    ------ value moved here
...
390 |                 let name_ptr = allocate_guest_pointer(caller, name_len);
    |                                                       ^^^^^^ value used here after move

error[E0382]: use of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:391:56
    |
377 |         move |mut caller: Caller<'_, T>, fields: u32, out_ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
390 |                 let name_ptr = allocate_guest_pointer(caller, name_len);
    |                                                       ------ value moved here
391 |                 let value_ptr = allocate_guest_pointer(caller, value_len);
    |                                                        ^^^^^^ value used here after move

error[E0382]: borrow of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:393:41
    |
377 |         move |mut caller: Caller<'_, T>, fields: u32, out_ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
391 |                 let value_ptr = allocate_guest_pointer(caller, value_len);
    |                                                        ------ value moved here
392 |
393 |                 let memory = memory_get(&mut caller).unwrap();
    |                                         ^^^^^^^^^^^ value borrowed here after move

For more information about this error, try `rustc --explain E0382`.
error: could not compile `wasi-http` due to 4 previous errors
```

Admittedly I'm a Rust newb, so if there is a way to resolve this, please let me know. I tried a few things but I couldn't see a way to do it.

File: crates/wasi-http/src/component_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/component_impl.rs]: switched to 4

File: crates/wasi-http/src/component_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/component_impl.rs]: removed, this was stale old code.

File: crates/wasi-http/src/component_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/component_impl.rs]: Done (here and below)

File: crates/wasi-http/Cargo.toml
Comment (brendandburns) [reply crates/wasi-http/Cargo.toml]: done.

File: crates/wasi-http/Cargo.toml
Comment (brendandburns) [reply crates/wasi-http/Cargo.toml]: done.

File: crates/wasi-http/Cargo.toml
Comment (brendandburns) [reply crates/wasi-http/Cargo.toml]: done.

File: crates/wasi-http/Cargo.toml
Comment (brendandburns) [reply crates/wasi-http/Cargo.toml]: done.

File: crates/wasi-http/Cargo.toml
Comment (brendandburns) [reply crates/wasi-http/Cargo.toml:2]: acknowledged.

File: crates/wasi-http/Cargo.toml
Comment (brendandburns) [reply crates/wasi-http/Cargo.toml]: done.

File: crates/wasi-http/src/component_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/component_impl.rs:129]: yeah, surprised me to, seems like some sort of aliasing in `wit-bindgen`

**Comment by brendandburns** [comment]
Comments addressed except the macro vs fn comment because I'm not clear how to resolve the compile error.

Please take another look.

File: crates/wasi-http/src/component_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/component_impl.rs]: I figured out how to fix this, so this is now fixed in the PR.

File: crates/wasi-http/src/http_impl.rs
Comment (Mossaka) [inline crates/wasi-http/src/http_impl.rs:27]:
```suggestion
            _ => return bail!("unknown method!"),
```

File: crates/wasi-http/src/lib.rs
Comment (Mossaka) [inline crates/wasi-http/src/lib.rs:4]: If you set `async: true` then you can use `reqwest::Client` instead of `reqwest::blocking::Client`.

Comment (pchickey) [reply crates/wasi-http/src/lib.rs:4]: Agreed - really, the only time we want to use the sync traits is to support hosts that arent running tokio. Since the reqwest blocking client is actually using tokio under the hood, using the sync trait means this library can't compose with the rest of the wasi libraries when run on tokio, but it also can't work for hosts that don't want to incur tokio, which is the worst of both worlds.

And once we make that change, I'd much prefer to drop down to just plain old hyper, because I don't think there's a ton of value in the reqwest async interface over the hyper primitives for this use case, and reqwest is responsible for at least half of the dependencies that need to be `cargo vet`ed in order to land this.

Comment (brendandburns) [reply crates/wasi-http/src/lib.rs:4]: I looked at hyper. It makes you do things like explicitly open a TCP connection before talking HTTP, it's ergonomics as a library are not great at first glance anyway.

I'm happy to do this as a later optimization, but I don't think that it impacts this PR, so I'd prefer to merge and iterate.

Comment (pchickey) [reply crates/wasi-http/src/lib.rs:4]: In order to land the PR I have to audit all unvetted dependencies for supply chain safety. These audits are very time consuming - I have to review the source of the entire crate. We can only accept these audits if they are performed by core team members.  I can't make any kind of exception for this PR, or breeze through things because reqwest will be swapped out soon, because our audits are imported to be part of the supply chain safety of other projects.

Right now this PR introduces 49 unvetted dependencies into the project https://github.com/bytecodealliance/wasmtime/actions/runs/4430843524/jobs/7773090032?pr=5929. That represents at least a full day of very tedious work for me. Many of those dependencies are pulled in exclusively by the reqwest library, in order to provide functionality we don't need. (Because of the use by other projects, vetting considers all dependencies for all features and targets, even though our project doesn't need them.)

Therefore, we need to iterate on this before landing the PR. Like I said, I can send a PR to your branch to help with this transformation, because my time would be better spent on making the changes that need to happen anyway vs auditing packages we don't care about.

Comment (brendandburns) [reply crates/wasi-http/src/lib.rs:4]: The need to reduce the scope of rust vet makes total sense to me. I will look into switching this to Hyper.

**Comment by brendandburns** [comment]
Ok, this was switched to hyper. I'm not sure it really saved you very many dependencies, but if there are other ways to reduce the dependency count, let me know. Hyper doesn't implement TLS, so I had to pull in `hyper_native_tls`

I think this is ready to go, let me know if there is a good way to add tests, it seems the other modules don't really have tests.

**Comment by brendandburns** [comment]
@pchickey 

Ok, I implemented timeouts. I think that the client side implementation is now basically feature complete. Ready for a more detailed review whenever you have time.

Thanks!

**Comment by brendandburns** [comment]
Oh, and I plan on starting the server side implementation in a different branch.

File: crates/wasi-http/Cargo.toml
Comment (eduardomourar) [inline crates/wasi-http/Cargo.toml:18]: i believe this was a typo
```suggestion
    http-body = "1.0.0-rc.2"
```

File: crates/wasi-http/Cargo.toml
Comment (eduardomourar) [inline crates/wasi-http/Cargo.toml:15]: we don't need the full set of features
```suggestion
    tokio = { version = "1", default-features = false, features = ["net", "rt-multi-thread", "time"] }
```

**Review by eduardomourar — CHANGES_REQUESTED** [review]
I have tested this with a real-life Wasm module and it works as expected.

**Review by eduardomourar — APPROVED** [review]

File: crates/wasi-http/src/streams_impl.rs
Comment (Mossaka) [inline crates/wasi-http/src/streams_impl.rs:11]: It doesn't seem like the `len` parameter has been used here to specify how much bytes to read. I feel like for an initial implementation this should be okay, but if the response body is too big to fit in host's memory then we probably need to use `BufRead` to read bytes into a buffer incrementally. 

I remember you added `read_into` function before, but I can't find it anymore.

**Review by Mossaka — APPROVED** [review]
mostly lgtm!

File: crates/wasi-http/src/streams_impl.rs
Comment (brendandburns) [reply crates/wasi-http/src/streams_impl.rs]: `read_into` was removed after feedback from @pchickey and others.

I fixed this function to actually implement read correctly.

**Comment by pchickey** [comment]
I left some fixes last night in https://github.com/brendandburns/wasmtime/pull/1 but it looks like they now conflict

**Comment by pchickey** [comment]
from my point of view we still need to fix 2 architectural issues before we can merge:
* wasmtime should be running in the tokio executor, rather than using block_on inside a Func. 
* rather than examples provided in C, we need to rewrite those to be test-programs written in Rust, because our CI environment doesn't have a C compiler installed.

I'll work on both of those things right now, and send them as PRs like the above:

**Comment by eduardomourar** [comment]
@pchickey , I believe you can borrow code from [here](https://github.com/eduardomourar/smithy-rs/blob/support-wasi-http/aws/sdk/integration-tests/webassembly/src/adapter/http_client.rs) for your Rust example.

**Comment by brendandburns** [comment]
@pchickey thanks for the PR, I merged it. Let me know if you want me to take on either of those tasks.

**Comment by brendandburns** [comment]
@pchickey unless you've already started on this, I can take care of those two items.

**Comment by brendandburns** [comment]
Ok, I switched the executor on the `Host` to use `tokio::Runtime::block_on` instead of `executors`.

I don't think it is required to build out the full async support in this first PR, but if that's really a blocker, I will.

I also added an example implemented in rust for use as a test or just another example.

Please take another look.

**Comment by pchickey** [comment]
I just finished writing basically the same rust example this morning but I didn't push it yet because I need to hook its build and execution into the `test-programs` machinery. I'll wrap that up tomorrow and push it to you.

And yeah lets just do async after we land this PR. We'll have to string async through everywhere, including into the wasmtime cli to run things under tokio when the http module is enabled.

**Comment by brendandburns** [comment]
@pchickey sounds good, I'm happy to delete my Rust example and just use yours if that is easier.

**Comment by brendandburns** [comment]
Cool. I squashed/merged that PR. Let me know if you need anything else from me.

**Comment by pchickey** [comment]
https://github.com/bytecodealliance/wasmtime/pull/6121 will need to land and merge to make the deny and vet CI steps pass. Then, we expect the rest of the steps will pass, since they did locally, but windows and mac can surprise me sometimes.

We will need to remove the contents of `crates/wasi-http/example/`, and at that point I think we are good to go, I'll make one pass with fresh eyes tomorrow.

**Comment by brendandburns** [comment]
* PRs merged (thanks!)
* example directory removed

I think this is ready to go.

**Comment by pchickey** [comment]
It looks like I missed the vet of some stuff I introduced in #2, I'll get those if you can fix the warnings I sloppily left in the test

**Review by pchickey — COMMENTED** [review]
This should fix the warnings

**Review by pchickey — APPROVED** [review]
All set!

**Comment by pchickey** [comment]
in order to make CI faster, we don't run every step until its time to go in the merge queue. https://github.com/brendandburns/wasmtime/pull/4 fixes the publishing issues.

File: crates/wasi-http/Cargo.toml
Comment (eduardomourar) [reply crates/wasi-http/Cargo.toml:2]: I know this is a little bit late, but can we change this package name to `wasmtime-wasi-http`, please?

**Comment by eduardomourar** [comment]
@pchickey , I know this is a little bit late, but can we change this package name to `wasmtime-wasi-http`, please? I feel that the current name there will eventually be used for the guest bindings and its high-level basic implementation.

**Comment by pchickey** [comment]
@eduardomourar I think that is a good idea but lets get things merged and then rename it - we have a few more days before the release gets cut.

**Comment by brendandburns** [comment]
Merged the PR. I will send a couple of follow on PRs (including the rename) this evening.

**Comment by eduardomourar** [comment]
Here is an initial stab at the guest binding: https://github.com/bytecodealliance/wasi/pull/69

**Comment by jeffparsons** [comment]
Is it easy/possible to play with these WASI module implementations from Rust as they land? E.g. if I copy a snapshot of the relevant wit files locally and then build a component using cargo-component and its adapter magic, could I then run that in Wasmtime and expect something to happen?

Just having a bit of trouble understanding what pieces already work together, and which need other pieces (new WASI target in rustc, completed component model support in Wasmtime) to land first.

**Comment by brendandburns** [comment]
@jeffparsons yes, it's pretty easy to play with the modules.

You need to:

Run `wit-bindgen rust wasmtime/crates/wasi-http/wit` this will generate the Rust client side bindings (you can substitute `wit-bindgen c ...` if you want the c bindings instead.

Once you have that, here's a small example program that you can compile:
```
use crate::proxy::types::new_fields;
use crate::proxy::types::new_outgoing_request;
use crate::proxy::types::MethodParam;
use crate::proxy::types::SchemeParam;

pub mod proxy;

fn request(method: MethodParam, scheme: Option<SchemeParam>, host: &str, path: &str) {
    let headers = new_fields(&[
        ("Content-type", "text/plain"),
        ("User-agent", "wasm32-wasi-rust"),
    ]);

    let req = new_outgoing_request(method, path, "", scheme, host, headers);
    let fut = crate::proxy::default_outgoing_http::handle(req, None);
    let res = crate::proxy::types::future_incoming_response_get(fut)
        .unwrap()
        .unwrap();
    let code = crate::proxy::types::incoming_response_status(res);
    let response_headers = crate::proxy::types::incoming_response_headers(res);
    let stream = crate::proxy::types::incoming_response_consume(res).unwrap();
    let body = crate::proxy::streams::read(stream, 60 * 1024).unwrap().0;

    println!("Status is {}", code);
    println!("Headers are:");
    let entries = crate::proxy::types::fields_entries(response_headers);
    for tuple in entries.iter() {
        println!("{}: {}", tuple.0, tuple.1);
    }
    println!("{}", String::from_utf8(body).unwrap());
}

fn main() {
    request(
        MethodParam::Get,
        Some(SchemeParam::Https),
        "postman-echo.com",
        "/get",
    );
    request(
        MethodParam::Post,
        Some(SchemeParam::Https),
        "postman-echo.com",
        "/post",
    );
    request(
        MethodParam::Put,
        Some(SchemeParam::Http),
        "postman-echo.com",
        "/put",
    );
}
```

File: crates/test-programs/build.rs
Comment (pchickey) [inline crates/test-programs/build.rs:1]:
```suggestion
#![allow(dead_code, unused_imports)]
```

**Comment by pchickey** [comment]
@jeffparsons we still need to land the preview 1 adapter and a component model implementation of the rest of wasi before this becomes usable from components. the preview 1 adapter means we don't have to worry about a new wasi target in the rust toolchain at this time.

**Comment by brendandburns** [comment]
@pchickey looks like the merge failed b/c libssl isn't installed on some of the test environments :(

How do you want to handle fixing this?

**Comment by pchickey** [comment]
I am a bit stumped on how to fix this.

To maximize linux libc compatibility, we perform our linux release builds on centos 7 (https://github.com/bytecodealliance/wasmtime/blob/main/ci/docker/x86_64-linux/Dockerfile). If we were to install the default libssl from centos 7 in the builder, it would mean all of our wasmtime linux users would need openssl 1.0.2 installed, which has been out of support since the start of 2020, so that isn't an acceptable solution.

I don't know enough about the linux and openssl ecosystem to know what minimum version is acceptable, but clearly it reduce our binary compatibility to take on this dep. I also don't know if there are any knobs we can tweak in the build to lazily load the openssl dylib, so users are opting in to needing it by passing the `--wasi-modules=experimental-wasi-http` flag.

Finally, I don't know if there are alternatives we can swap in for needing openssl on linux. I am aware that `rustls` exists, but I don't know if it is a complete alternative. I'll ask some of my colleagues who have spent more time in this space.

**Comment by rvolosatovs** [comment]
> I am a bit stumped on how to fix this.
> 
> To maximize linux libc compatibility, we perform our linux release builds on centos 7 ([`main`/ci/docker/x86_64-linux/Dockerfile](https://github.com/bytecodealliance/wasmtime/blob/main/ci/docker/x86_64-linux/Dockerfile?rgh-link-date=2023-03-31T17%3A57%3A37Z)). If we were to install the default libssl from centos 7 in the builder, it would mean all of our wasmtime linux users would need openssl 1.0.2 installed, which has been out of support since the start of 2020, so that isn't an acceptable solution.
> 
> I don't know enough about the linux and openssl ecosystem to know what minimum version is acceptable, but clearly it reduce our binary compatibility to take on this dep. I also don't know if there are any knobs we can tweak in the build to lazily load the openssl dylib, so users are opting in to needing it by passing the `--wasi-modules=experimental-wasi-http` flag.
> 
> Finally, I don't know if there are alternatives we can swap in for needing openssl on linux. I am aware that `rustls` exists, but I don't know if it is a complete alternative. I'll ask some of my colleagues who have spent more time in this space.

FWIW, we have used `rustls` in https://github.com/enarx/enarx to provide transparent TLS to guests by wrapping the wasmtime WASI TCP traits and it was perfectly sufficient for that particular use case, which seems quite related.

Depending on `rustls` would be best for both compatibility and simplicity of development. Definitely would vote for depending on `rustls` instead of `openssl`

This would also make (perhaps eventually) static builds much easier

**Comment by pchickey** [comment]
Thanks, I asked around and we don't use rustls inside fastly for reasons that aren't relevant to this case. @brendandburns are you up for switching things to use rustls? I can do whatever supply chain audits that requires as soon as we know which crates it will take.

I guess it is a good time to rename the `wasi-http` crate to `wasmtime-wasi-http` in this PR since we'll be squeaking in before the release is cut next week.

**Comment by pchickey** [comment]
I was just told we use rustls in https://github.com/fastly/viceroy and it has worked great for that use case.

**Comment by brendandburns** [comment]
I will try switching to `rustls` I believe that it should be a pretty easy switch. I have a commit for the rename queued up also, so I will add that to this PR too.

File: .github/workflows/main.yml
Comment (rvolosatovs) [inline .github/workflows/main.yml:397]: Shouldn't need SSL anymore, right?

Comment (brendandburns) [reply .github/workflows/main.yml]: done.

**Comment by pchickey** [comment]
I'm not a license lawyer at all but it appears the config our team selected for cargo-deny doesn't permit the use of `webpki-roots`:
https://github.com/bytecodealliance/wasmtime/actions/runs/4579627092/jobs/8087593223?pr=5929#step:5:160
Could we swap it out to use https://github.com/rustls/rustls-native-certs instead?

**Comment by tschneidereit** [comment]
If our cargo deny config doesn't accept MPL I'd argue we should change it. The MPL 2.0 is one of the most well designed and carefully legally vetted licenses out there, and has all the key properties we need, in particular only per-file copyleft and a patent grant

File: crates/wasi-http/src/http_impl.rs
Comment (eduardomourar) [inline crates/wasi-http/src/http_impl.rs:201]: we should actually be returning the future incoming response id here and not the response id

Comment (brendandburns) [reply crates/wasi-http/src/http_impl.rs:201]: Future's aren't implemented (yet) in this PR so the response_id is the same as the "future" id. If you look at the future-incoming-response-get in component_impl.rs, you'll see that it just returns the passed in future as the response id.

https://github.com/bytecodealliance/wasmtime/pull/5929/files/f02db0f19f99eeb82c369b511cb56743f31b7f60#diff-902e7d68ea9ed72ae165cb94771bad9528553b6b1a04661f731bbe40460ff1beR215

Eventually we will implement futures and this will change, but not in this PR.

**Review by pchickey — APPROVED** [review]
Thank you for making these changes so quickly. I believe the supply chain and licensing stuff is just a review away from merging, once those go in I'll need you to merge main in, and it should work.

File: crates/wasi-http/src/http_impl.rs
Comment (eduardomourar) [reply crates/wasi-http/src/http_impl.rs:201]: after this is merged, i will submit a pr that will at least prevent the host from crashing if the emulated future is assumed by the guest like here: https://github.com/bytecodealliance/jco/blob/2b3c9cd7839b3728f74af2cbf6722cf09287f956/packages/preview2-shim/lib/http/wasi-http.js#L263-L296

**Comment by pchickey** [comment]
@brendandburns please merge `main` and this should pass CI, fingers crossed

**Comment by brendandburns** [comment]
@pchickey thanks, rebased on `main` this looks like it's ready to go (again :))

**Comment by brendandburns** [comment]
@pchickey I missed the rename in `scripts/publish.rs` so this got bounced out of the queue. 

This is now fixed, hopefully it will pass the merge queue this time...

**Comment by pchickey** [comment]
Unfortunately, it appears `ring` does not support riscv64. I think we will have to juggle some cargo feature flags around to disable including the `wasmtime-wasi-http` crate into `wasmtime-cli` on that platform.

Also, I just learned that you can use the magic string `prtest:full` in a commit message, and github will run the complete CI instead of just the automatically-estimated subset. That will at least get these issues to show up more eagerly, rather than waiting to enter the main merge queue.

**Comment by brendandburns** [comment]
@pchickey ok, I will see about flipping the feature flags the right way (and thanks for the tip on the prtest:full, that will definitely enable quicker iteration.

**Comment by brendandburns** [comment]
@pchickey I conditionalized tls to disable it on `riscv64gc`

I tried the `prtest:full` thing but it didn't seem to work for me, maybe I did it wrong...

Anyhow, perhaps we can give this another try? If there's a way for me to trigger the full test runs w/o you adding it to the merge queue, let me know what I need to change in my commit message.

Thanks!

**Comment by brendandburns** [comment]
cc @pchickey 

Ok, next issue hit:

The `wasmtime::component::bindgen!` macro expects a folder named `wit` in the same directory as the `Cargo.toml`

The challenge is that if we want to pull the `wasi-http` repository in as a git submodule, then the directory structure is messed up, so I used a symbolic link to link `wit` -> `wasi-http/wit` but it looks like those symbolic links don't work properly for Windows.

I can get rid of the git submodule, and just check in the `wit` files directly into this repo, it will make it a little more annoying to keep the wit files in-sync but it's do-able.

Let me know what you think.


`

**Comment by brendandburns** [comment]
It looks like _maybe_ setting `git config --global core.symlinks true` in our builds would fix things? But I'm not certain where those builds are configured, I can't seem to find them in `main.yml`

**Comment by cfallin** [comment]
Is it possible to modify the macro to have a configurable path? I think we'd probably prefer that over most any directory-structure or build-time hack (symlinks, copying files, or the like).

**Comment by eduardomourar** [comment]
I believe the macro does accept a path property and you can check it here: https://github.com/bytecodealliance/wasmtime/blob/main/crates/component-macro/tests/codegen.rs

**Comment by brendandburns** [comment]
@eduardomourar thanks for the pointer (I don't think that's documented anywhere :() I updated this to use `path` and I removed the symlink.

*crosses fingers*

**Comment by brendandburns** [comment]
@pchickey This passed complete CI, but it needed a rebase. I'm pretty sure that it will pass the merge queue this time.

Thanks for the patience!

File: crates/wasi-http/Cargo.toml
Comment (pchickey) [inline crates/wasi-http/Cargo.toml:22]:
```suggestion
# The `ring` crate, used to implement TLS, does not build on riscv64 or s390x
[target.'cfg(not(any(target_arch = "riscv64", target_arch = "s390x")))'.dependencies]
```

**Review by pchickey — APPROVED** [review]

**Comment by brendandburns** [comment]
🥂 🎊 🥂 

Thanks for the patience here, I'll send in some follow-on PRs.

**Pull request merged**
